src/Evolution/Systems/RadiationTransport/M1Grey/M1Closure.cpp

```diff
@@ -0,0 +1,284 @@
+    const tnsr::ii<DataVector, 3, Frame::Inertial>& spatial_metric,
+    const tnsr::II<DataVector, 3, Frame::Inertial>&
+        inv_spatial_metric) noexcept {
+  // Small number used to avoid divisions by zero
+  static constexpr double avoid_divisions_by_zero = 1.e-150;
+  // Below small_velocity, we use the v=0 closure,
+  // and we do not differentiate between fluid/inertial frames
+  static constexpr double small_velocity = 1.e-15;
+  Variables<tmpl::list<
+      hydro::Tags::LorentzFactorSquared<DataVector>, MomentumSquared,
+      MomentumUp,
+      hydro::Tags::SpatialVelocityOneForm<DataVector, 3, Frame::Inertial>,
+      hydro::Tags::SpatialVelocitySquared<DataVector>>>
+      temp_closure_tensors(get(energy_density).size());
+
+  // The main calculation needed for the M1 closure is to find the
+  // roots of J^2 zeta^2 = H^a H_a, with J the fluid-frame energy density
+  // and H^a the fluid-frame momentum density (0th and 1st moments).
+
+  // We first compute various fluid quantities and contractions
+  // with inertial moments needed even if v^2 is small
+  auto& w_sqr =
+      get<hydro::Tags::LorentzFactorSquared<DataVector>>(temp_closure_tensors);
+  get(w_sqr) = square(get(fluid_lorentz_factor));
+  auto& v_sqr = get<hydro::Tags::SpatialVelocitySquared<DataVector>>(
+      temp_closure_tensors);
+  get(v_sqr) = 1. - 1. / get(w_sqr);
+  // S^i, the neutrino momentum tensor
+  auto& s_M = get<MomentumUp>(temp_closure_tensors);
+  raise_or_lower_index(make_not_null(&s_M), momentum_density,
+                       inv_spatial_metric);
+  // S^i S_i
+  auto& s_sqr = get<MomentumSquared>(temp_closure_tensors);
+  dot_product(make_not_null(&s_sqr), s_M, momentum_density);
+  // v_i, the spatial velocity one-form of the fluid
+  auto& v_m =
+      get<hydro::Tags::SpatialVelocityOneForm<DataVector, 3, Frame::Inertial>>(
+          temp_closure_tensors);
+  raise_or_lower_index(make_not_null(&v_m), fluid_velocity, spatial_metric);
+
+  // Allocate memory for H^i
+  std::vector<double> H_M(3, 0.);
+  // Loop over points
+  for (size_t s = 0; s < v_sqr.size(); ++s) {
+    const double& v_sqr_pt = get(v_sqr)[s];
+    const double& w_sqr_pt = get(w_sqr)[s];
+    const double& e_pt = get(energy_density)[s];
+    const double& s_sqr_pt = std::max(get(s_sqr)[s], avoid_divisions_by_zero);
+    // Ignore complex closure calculations
+    // if the fluid velocity is very small
+    if (v_sqr_pt < small_velocity) {
+      // Minerbo closure assuming v=0 (see definition of
+      // minerbo_closure_function)
+      const double zeta = sqrt(s_sqr_pt) / e_pt;
+      const double chi = minerbo_closure_function(zeta);
+      const double d_thin_e_pt_over_s_sqr = (1.5 * chi - 0.5) * e_pt / s_sqr_pt;
+      const double d_thick_e_pt_over_3 = (1.5 - 1.5 * chi) * e_pt / 3.;
+
+      get(*closure_factor)[s] = zeta;
+      get(*comoving_energy_density)[s] = e_pt;
+      get(*comoving_momentum_density_normal)[s] = 0.;
+      for (size_t i = 0; i < 3; i++) {
```

Comment:
define `constexpr size_t spatial_dim = 3;` somewhere in the beginning of the function and use `spatial_dim` in the loops (assuming 3 is the spatial dimension)

Reply:
I am changing 3->spatial_dim in the loops, but leaving the '3' in the templated definition of tensors, e.g.
auto& v_m =
      get<hydro::Tags::SpatialVelocityOneForm<DataVector, 3, Frame::Inertial>>(
          temp_closure_tensors);

Is that ok? Changing the 3's in the tensors would require also changing them for the tensors provided as inputs, which may require templating the class on the dimension?

Reply:
yes that is ok